qiskit/circuit/quantumcircuit.py

```diff
@@ -1494,15 +1494,23 @@ def decompose(
 
         Returns:
             QuantumCircuit: a circuit one level decomposed
+
+        Raises:
+            CircuitError: if gates_to_decompose exist with more than one repeat.
         """
         # pylint: disable=cyclic-import
         from qiskit.transpiler.passes.basis.decompose import Decompose
         from qiskit.converters.circuit_to_dag import circuit_to_dag
         from qiskit.converters.dag_to_circuit import dag_to_circuit
 
-        pass_ = Decompose(gates_to_decompose=gates_to_decompose, reps=reps)
-        decomposed_dag = pass_.run(dag=circuit_to_dag(self))
-        return dag_to_circuit(decomposed_dag)
+        pass_ = Decompose(gates_to_decompose)
+        dag = circuit_to_dag(self)
+        if gates_to_decompose is not None and reps > 1:
+            raise CircuitError("gates_to_decompose won't work with other than one repeat.")
```

Comment:
I think this can work for more than one repetition, for example assume you have a circuit consisting of a controlled phase (`cp`) gate and set `gates_to_decompose=["cp", "p"]`. Then in the first decomposition, the gate will be decomposed into phase gates (`p`) and CX, and in the second repetitions the phase gates are decomposed into `U` gates.

Reply:
@Cryoris  my idea is the same, but do the repetition without this worry, and can decompose specifc label gate specific time, so it won't decompose another gate that have the same ``op.name`` gate.